### DISCUSSION TITLE
Error on building static page

### DISCUSSION THREAD

**Discussion by coldfeudal** [vercel/next.js#37711] [Help] [answered]
How do I build with `getStaticProps` on a static page? My api is on the same next project, but I can't build with it running in dev as it throws `EPERM: operation not permitted`

```js
// Page: "/pages/catalog.js"

const Catalog = ({ dbProducts, dbCategories }) => {

...


export const getStaticProps = async () => {
    const res1 = await fetch(process.env.DOMAIN_NAME + "api/getNProducts/24")
    const dbProducts = await res1.json()

    const res2 = await fetch(process.env.DOMAIN_NAME + "api/getCategories")
    const dbCategories = await res2.json()

    return {
        props: {
            dbProducts,
            dbCategories
        }
    }
}

export default Catalog
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Please do think about it like this, in order for `api/routes` to exist, you must have a running server, but at build time, there's no such thing, so it is impossible to reach any `api/routes`, during this phase. 

ISR might lead you to believe otherwise, same with development mode, where `getStaticProps` runs aggressively, or as needed for ISR. In plain terms, `getStaticProps` runs only at build time, unless triggered by ISR.

What the documentation recommends, and it is a very good practice, is to directly reach for the service you want to query from.

I always use this example:

- A util/helper function to reach my 3rd party , in this case codewars

```ts
// lib/codewars.ts
export const getCodeWarsUser = (user) =>
  axios
    .get(`https://www.codewars.com/api/v1/users/${user}`)
    .then(({ data }) => data);
```

- Then in getStaticProps, inside a page

```tsx
// pages/index.ts
export async function getStaticProps() {
  const codewars = await getCodeWarsUser('some-user');

  return {
    props: { codewars },
    revalidate: 10,
  };
}
```

- And also in `api/codewars`, so I can reach it from the client, or any other http client really

```ts
// api/codewars.ts
const codewars = async (req, res) => {
  const data = await getCodeWarsUser('some-user');

  return res.send(data);
};

export default codewars;
```

The idea is, to design how you reach the service you want to get data from, in such a way that you can access it both from `api/routes` and from `getStaticProps`, and `getServerSideProps`.

Another way to look at this is, all, `api/routes`, `getStaticProps` and `getServerSideProps` do not run in the browser, they have the capability of running in the server, so why would you already being in the server, reach out to a sibling function, that is `api/route` from 
`getStaticProps` or `getServerSideProps`, if you could just do the same thing you'd do inside `api/routes`.

If you are concerned about bundle sizes, or leaking secrets to the client, you can use this tool https://next-code-elimination.vercel.app/ to see what parts of the page code are bundled for the client.

### A quirk

I hope it is clear why, even if you use `process.env.DOMAIN_NAME` to create a full path, during build time, it is impossible to reach the `api/routes`.

I'd also recommend avoiding this for `getServerSideProps` just connect to the API you want to consume from directly! 

You are already running server side, another http transaction is not necessary at all.

**Reply by coldfeudal** [reply]
So basically there is no way to use `getStaticProps` to get data from api on the same project? Trying to get data from db once on build. Am I right assuming I will need a separate server for that?

**Reply by icyJoseph** [reply]
> So basically there is no way to use getStaticProps to get data from api on the same project? 

By definition, during build time, there's no `api/routes` live.

> Trying to get data from db once on build. Am I right assuming I will need a separate server for that?

**Incorrect**. You can connect to the DB during build time. You don't need a separate server at all (of course, assuming the DB is running elsewhere as an independent thing, reachable by URL)

Consider, https://next-code-elimination.vercel.app/s/3b9N1Z66L, it can just connect to the DB from `getStaticProps`, or `getServerSideProps`. 

Again, think about it, `getStaticProps` is running on a node environment, so any connections, queries and such you need to get data from your DB, you can directly do inside this function.

`getStaticProps` or `getServerSideProps` are a sort of specialised `api/route`, that directly pipe props into a page, at build-time or runtime.

I see quite a few people get stumped by this, of all videos I've watched on this matter, I think this video gives a very clear and concise explanation, exactly at this point: https://youtu.be/MxlmfI9IxVs?t=523
